### DISCUSSION TITLE
Sync background thread to main thread in .NET Core console app?

### DISCUSSION THREAD

**Discussion by artemiusgreat** [dotnet/reactive#1778] [Q&A] [answered]
Trying various combinations of these schedulers with the code below to achieve switching of synchronization context like it's described in the [article](https://devblogs.microsoft.com/pfxteam/await-synchronizationcontext-and-console-apps/). 

**Class the will send notifications from background thread**

```
public class RxProcessor
{
  public Subject<dynamic> observable = null;

  public RxProcessor(Action action)
  {
    var ctx = SynchronizationContext.Current ?? new();
    var scheduler = CurrentThreadScheduler.Instance;

    observable = new();
    observable.ObserveOn(scheduler).Subscribe(o => action());  // how to break uot from background to main thread?
  }
}
``` 

Create background threads and start sending notifications 

```
public static int Log(string txt)
{
  var message = $"{txt} - { Thread.CurrentThread.ManagedThreadId }";
  Console.WriteLine(message);
  return Thread.CurrentThread.ManagedThreadId;
}

Log("Main Thread");
var rx = new RxProcessor(() => Log("RX Thread"));
Parallel.For(0, 10, o => rx.observable.OnNext("Demo"));
```

I would like to see the same thread ID at the moment of subscription and observation. 

```
Main Thread - 1
RX Thread - 1
RX Thread - 1
RX Thread - 1
RX Thread - 1
RX Thread - 1
RX Thread - 1
RX Thread - 1
RX Thread - 1
RX Thread - 1
RX Thread - 1
```

Unfortunately, the observation thread is different from the main subscription thread. 

```
Main Thread - 1
RX Thread - 13
RX Thread - 13
RX Thread - 13
RX Thread - 13
RX Thread - 13
RX Thread - 13
RX Thread - 13
RX Thread - 13
RX Thread - 13
RX Thread - 13
```

Is it possible to somehow always receive notifications in the main thread even though `OnNext()` was called in the background thread?

**Comment by cabauman** [comment, marked as answer]
Working with a console app is a bit different than a GUI app in this regard. It doesn't have an event loop for Rx to work with. A solution [shared here](https://stackoverflow.com/a/37872567/5984310) uses an EventLoopScheduler to fill in as a "main thread." Maybe that will work for your use case?

**Reply by artemiusgreat** [reply]
Thank you, it's close to what I'd like to achieve, the only thing is that this approach doesn't provide thread safety because any other call on the default thread pool, including `async`, will try to access variables from another thread.
